### PR TITLE
Merge otherlibs/dynlink/Makefile into the root Makefile

### PR DESCRIPTION
**Pull request by shindere**
This PR continues the build system unification work started in #11243, #11248, #11268, #11420, #11675, #12198, #12321, #12586, #12616, #12706 and #13048.

The present PR merges `otherlibs/dynlink/Makefile` into the root `Makefile` and consists of four commits. The three first ones have the property of not altering the installed artefacts at all. The only commit that does so is the last one, which is the core of the PR, and the only thing it changes to the installed artifacts is the checksums, which is due to the fact that the compilation of the OCaml sources is done from the top-level of the source tree rather than from the `otherlibs/dynlink` directory as before.

The first commit introduces the private `OTHERLIBS` build variable to keep track of which other libraries are still built by `otherlibs/Makefile`. This variable is initially equal to `OTHERLIBRARIES` (the list of other libraries to build) but this list will become shorter andshohrter as more and more other libraries will be built directly by the root Makefile. Once all the other libraries are built by the root Makefile this variable is expected to become empty and to finally vanish.

The second commit delegates some file copying which was previously done during the build to the configure script.

THe third (and last preparatory) commit extends the general framework defined in `Makefile.common` to make sure the `.cmx` files of a native library get copied to the directory where its `.cmxa` file resides, as this is necessary when linking with dynlink's native version. As the copy happens also for all the other libraries built by the framework, the `clean` rule of `compilerlibs` had to be adjusted to clean the newly copied artefacts.

Finally, the PR's last commit is its core, which is the actual merge of `otherlibs/dynlink/Makefile` into the root `Makefile`. As usually, this goes hand in hand with `otherlibs/dynlink/.depend` being merged into the root `.depend`.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1559,7 +1586,13 @@ clean::
 
 # Dependencies
 
-subdirs = stdlib $(addprefix otherlibs/, $(ALL_OTHERLIBS))
+# The following definition duplicates the otherlibs/ prefix but this
+# will go away with the merge of the sub  makefiles
```
Comment (ghost) [inline Makefile]:
```suggestion
# will go away with the merge of the sub makefiles.
```

**Comment by shindere** [comment]
Fixed, thanks!

**Comment by shindere** [comment]
A few updates:

The PR is now ready to be reviewed, after a bit of work,
part of which having been done offline with @dra27.

To elaborate on the change of checksums, this is actually due
to the fact that this PR changes the file names that are
stored in locations. Instead of having jsut file names as we
did before, we now have complete file paths relative to the
top level directory of the source tree. This can be seen in the
changes to the test reference files.

Also, thePR now begins with a commit that simplifies the `C#` test
by not mentionning the dynlink files explicitly. THey are actually
already linked so they were linked twice, which caused a problem when used
in conjunction with `-linkall`. This is @dra27's discovery.

The commit which had been added to copy `.cmx` files systematically has
finally been remove in favour of a more ad hoc solution, which is closer
tohwat was done before.

**Review by dra27 — APPROVED** [review]
Reviewed offline (outcome in the most recent force push) - LGTM, thanks, @shindere

**Comment by shindere** [comment]
Plenty of thanks @ëra27 for your super supportive review.

**Pull request merged**